feat(core): added clang-format

iraizo opened the pull request:
<!-- First of all, THANK YOU for your contribution. -->

## Changes Proposed:
-  Added .clang-format file
-  

## Issues Addressed:
<!-- If your fix has a relating issue, link it below -->
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/6847

## SOURCE:
<!-- If you can, include a source that can strengthen your claim -->
[clang format documentation](https://clang.llvm.org/docs/ClangFormat.html)

## Tests Performed:
<!-- Does it build without errors? Did you test in-game? What did you test? On which OS did you test? Describe any other tests performed -->
- ![](https://i.imgur.com/qIDOPn5.png)
- 


## How to Test the Changes:
<!-- Describe in a detailed step-by-step order how to test the changes -->

1. `clang-format style=file *.cpp *.h`
2.
3.

## Known Issues and TODO List:
<!-- Is there anything else left to do after this PR? -->

- [ ] Format the whole project
- [ ]

<!-- If you intend to contribute repeatedly to our project, it is a good idea to join our discord channel. We set ranks for our contributors and give them access to special resources or knowledge: https://discord.com/invite/DasJqPba)
     Do not remove the instructions below about testing, they will help users to test your PR -->
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

iraizo commented:
Obviously there can be more added to the format but i have made minimal changes, i recommend trying out some styles using https://zed0.co.uk/clang-format-configurator/

Kitzunu commented:
As long as it follows https://www.azerothcore.org/wiki/cpp-code-standards

iraizo commented:
> As long as it follows https://www.azerothcore.org/wiki/cpp-code-standards

Look at the screenshot for the format, brackets and whitespaces have been added with some additional rules.

Winfidonarleyan commented:
@iraizo configure this to https://www.azerothcore.org/wiki/cpp-code-standards

iraizo commented:
> @iraizo configure this to https://www.azerothcore.org/wiki/cpp-code-standards

Im not sure how I should implement the other coding standards discussed on that page Baded on my knowledge or if it's even possible

FrancescoBorzi commented:
> > @iraizo configure this to https://www.azerothcore.org/wiki/cpp-code-standards
> 
> Im not sure how I should implement the other coding standards discussed on that page Baded on my knowledge or if it's even possible

as long as they don't go **against**, it's fine. Not all conventions can be automated I guess

iraizo commented:
> > > @iraizo configure this to https://www.azerothcore.org/wiki/cpp-code-standards
> 
> > 
> 
> > Im not sure how I should implement the other coding standards discussed on that page Baded on my knowledge or if it's even possible
> 
> 
> 
> as long as they don't go **against**, it's fine. Not all conventions can be automated I guess

Yeah I rewrote it a bit and added some rulesets to it that makes code easier to read, but you're correct clang-format can't automate that all, naming conventions can be detected using clang-tidy but to my knowledge can't be automated

iraizo commented:
This is the current style

![](https://i.imgur.com/KkBla6K.png)

iraizo commented:
@FrancescoBorzi Please review this

FrancescoBorzi commented:
> This is the current style
> 
> ![](https://camo.githubusercontent.com/269ed8226abc485bbcd7a676a96c5654cf20fcecd013bf1e2fe534182c690efb/68747470733a2f2f692e696d6775722e636f6d2f4b6b426c61364b2e706e67)

why brackets of loops are not in line with the main loop statement?

I mean, it should be like this:

```cpp
while()
{
    // some code
}
```

instead of this:

```cpp
while()
  {
      // some code
  }
```

iraizo commented:
The issue is probaly because of the snippet below, let me check.

```
TabWidth: '4'
UseTab: ForIndentation
```

iraizo commented:
The `BreakBeforeBraces` template from GNU does that.

iraizo commented:
![](https://i.imgur.com/cKtx33e.png)

FrancescoBorzi commented:
thanks @iraizo , I guess it should be fine now

iraizo commented:
Yep!

Kitzunu commented:
> The issue is probaly because of the snippet below, let me check.
> 
> ```
> TabWidth: '4'
> UseTab: ForIndentation
> ```

Not tabs, 4 spaces.

iraizo commented:
> > The issue is probaly because of the snippet below, let me check.
> > ```
> > TabWidth: '4'
> > UseTab: ForIndentation
> > ```
> 
> Not tabs, 4 spaces.

It would use 1 tab what is equal to 4 spaces, im not sure whats wrong here

iraizo commented:
Actually it does seem to indent right before the bracewrapping overwrites the indention so i guess i will remove it before that

iraizo commented:
@Kitzunu Actually this has already been removed in another commit and has been ruled as `UseTab: Never` to atleast not make the formatter use tabs and hopefully also remove them

iraizo commented:
Things to still add:
- [ ]  Add a whitespace between the if and the condition
- [ ]  Add the 4 whitespaces

iraizo commented:
The config now automates the checks of:
- [x] whitespaces
- [x] brackets
- [ ] random numbers & constants | Not possible to automate
- [ ] Enumerations vs. define | Possibly a job for clang-tidy
- [ ] Standard prefixes for constants | Possibly a job for clang-tidy
- [ ] Naming of variables and functions | A perfect job for clang-tidy
- [ ] WorldObjects | Not possible to automate (without custom tools)
- [ ] Defining const variables | A perfect job for clang-tidy
- [ ] Static variables | A job for clang-tidy
- [x] Pointer alignment 

All other checks possible by clang-tidy could be made by using the `-fix` flag to automate them, but not every function has that and will not be added to this PR, maybe in the future.

This is ready to be reviewed

FrancescoBorzi commented on .clang-format:
```diff
@@ -0,0 +1,32 @@
+SpacesInCStyleCastParentheses: 'false'
+SpacesInContainerLiterals: 'false'
+SpacesInParentheses: 'false'
+Standard: Cpp11
```
shouldn't it be `Cpp17` ? as we are using C++17

iraizo replied:
Even tho i am using clang12 and the docs say cpp17 and even cpp20 is supported it will error as soon as im trying to format anything.

FrancescoBorzi replied:
I don't get it o_O

iraizo replied:
<img src="https://i.imgur.com/5k471NQ.png">

Winfidonarleyan replied:
![image](https://user-images.githubusercontent.com/18329778/125205407-441dfb80-e2ac-11eb-98ec-060d91a4481b.png)

Kitzunu replied:
According to https://zed0.co.uk/clang-format-configurator/ which Iraizo linked, this tool only has 

```
Format compatible with this standard, e.g. use A<A<int> > instead of A<A<int>> for LS_Cpp03.

Possible values:

LS_Cpp03 (in configuration: Cpp03) Use C++03-compatible syntax.

LS_Cpp11 (in configuration: Cpp11) Use features of C++11, C++14 and C++1z (e.g. A<A<int>> instead of A<A<int> >).

LS_Auto (in configuration: Auto) Automatic detection based on the input.
```

Kitzunu replied:
So maybe this tool is not good

Winfidonarleyan commented on .clang-format:
```diff
@@ -0,0 +1,34 @@
+SpacesInCStyleCastParentheses: 'false'
+SpacesInContainerLiterals: 'false'
+SpacesInParentheses: 'false'
+Standard: Cpp11
```
```suggestion
Standard: Latest
```

iraizo replied:
This works instead of specifying a certain version!

Winfidonarleyan replied:
Hah, just check doc :D

iraizo replied:
I did check it but i didnt actually see the Latest option and just read the `Cpp...` options, my bad ^^

iraizo commented:
Now the issue with this format still relies on how clang-format does line breaks on certain lines, see commit https://github.com/azerothcore/azerothcore-wotlk/pull/6872/commits/2779246025da1c8f3ed0a677ff69eb8bc9ab04e3#diff-b7f75e44a164603aa5c71dd2684d30556a5149ff490b9eb092ee5e09c5374d12R578 I've tried to fix this by this commit https://github.com/azerothcore/azerothcore-wotlk/pull/6848/commits/b56fcbdc9cbe66ec79424cf4358cfb357e62db04
but this seems to fix some of those line breaks, especially function declarations but causes some on function calls.

FrancescoBorzi commented:
I guess we can merge this and if there are leftovers we can always address them later.

Thanks @iraizo !

iraizo commented:
Sure!

iraizo commented:
This is ready to be merged after my next commit, see https://github.com/azerothcore/azerothcore-wotlk/pull/6872#issuecomment-879363919

iraizo commented:
Check comment https://github.com/azerothcore/azerothcore-wotlk/pull/6872#issuecomment-879366836 for what changed.
Ready to be merged

iraizo commented:
@FrancescoBorzi Should issues be created for the tasks described in this PR or how are they handled?

FrancescoBorzi commented:
> @FrancescoBorzi Should issues be created for the tasks described in this PR or how are they handled?

what do you mean?

iraizo commented:
> > @FrancescoBorzi Should issues be created for the tasks described in this PR or how are they handled?
> 
> what do you mean?

Im talking "about the Known Issues and TODO List"
